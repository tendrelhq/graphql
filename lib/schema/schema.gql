enum AuthenticationProvider {
  legacy
  clerk
}

enum LanguageCode {
  en
  es
}

type Language {
  id: ID!
  code: LanguageCode!
  name: Name!
  name_id: ID!
}

type Name {
  id: ID!
  language: Language!
  language_id: ID!
  value: String!
}

type Tag {
  id: ID!
  name: Name!
  name_id: ID!
  type: String!
  parent: Tag
  parent_id: ID
}

type Customer {
  id: ID!
  active: Boolean!
  name: Name!
  name_id: ID!
  defaultLanguage: Language!
  default_language_id: ID!
}

type Location {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  name: Name!
  name_id: ID!
  parent: Location
  parent_id: ID
  scan_code: ID
  site: Location!
  site_id: ID!
  tags: [Tag!]!

  children(options: LocationsQueryOptions): [Location!]!
}

type User {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  authentication_provider: Tag
  authentication_provider_id: ID
  language: Language!
  language_id: ID!
  name: String!
  tags: [Tag!]!
}

type Worker {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  language: Language!
  language_id: ID!
  role: Tag!
  role_id: ID!
  scan_code: String
  tags: [Tag!]!
  user: User!
  user_id: ID!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  languages: [Language!]!
  locations(customerId: ID!, options: LocationsQueryOptions): [Location!]!
  location(id: ID!): Location!
  roles: [Tag!]!
  workers(customerId: ID!, options: WorkersQueryOptions): [Worker!]!
}

input LocationsQueryOptions {
  cornerstone: Boolean
  site: Boolean
}

input WorkersQueryOptions {
  active: Boolean
  site: ID
}

type Mutation {
  # Location CRUD
  createLocation(input: CreateLocationInput!): Location!
  updateLocation(input: UpdateLocationInput!): Location!

  # User
  ## CRUD
  createUser(input: CreateUserInput!): User!
  updateUser(input: UpdateUserInput!): User!
  ## Helpers
  acceptInvite(input: AcceptInviteInput!): User!

  # Worker CRUD
  createWorker(input: CreateWorkerInput!): Worker!
  updateWorker(input: UpdateWorkerInput!): Worker!
}

input CreateLocationInput {
  id: ID
  active: Boolean!
  name: CreateNameInput!
  org_id: ID!
  scan_code: ID
}

input UpdateLocationInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  name: UpdateNameInput
  scan_code: ID
}

input CreateUserInput {
  id: ID
  active: Boolean!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input UpdateUserInput {
  id: ID!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input CreateWorkerInput {
  id: ID
  active: Boolean!
  language_id: ID!
  org_id: ID!
  role_id: ID!
  scan_code: ID
  user_id: ID!
}

input UpdateWorkerInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  role_id: ID!
  scan_code: ID
}

input CreateNameInput {
  id: ID
  active: Boolean!
  language_id: ID!
  value: String!
}

input UpdateNameInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  value: String!
}

input AcceptInviteInput {
  id: ID!
  authentication_identity: ID!
}