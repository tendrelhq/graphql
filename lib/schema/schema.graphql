enum Language {
  en
  es
}

type Tag {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

type Customer {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

type Location {
  id: ID!
  customerId: ID!
  name: String!
  parentId: ID
  tags: [Tag!]!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  locations(customerId: ID!, parentId: ID): [Location!]!
  sites(customerId: ID!): [Location!]!
}

type Mutation {
  createCustomer(input: CreateCustomerInput!): ID!
  updateCustomer(input: UpdateCustomerInput!): ID!
  deleteCustomer(id: ID!): ID!
  createLocation(input: CreateLocationInput!): ID!
  updateLocation(input: UpdateLocationInput!): ID!
  deleteLocation(id: ID!): ID!
}

input CreateTagInput {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

input UpdateTagInput {
  id: ID!
  name: String
}

input CreateCustomerInput {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

input UpdateCustomerInput {
  id: ID!
  name: String
  defaultLanguage: Language
}

input CreateLocationInput {
  id: ID!
  customerId: ID!
  name: String!
  tags: [CreateTagInput!]!
}

input UpdateLocationInput {
  id: ID!
  name: String
  tags: [UpdateTagInput!]
}
