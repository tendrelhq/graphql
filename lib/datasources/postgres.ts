import type { Request } from "express";
import z from "myzod";
import postgres, { type Fragment } from "postgres";
import { makeActiveLoader } from "./activatable";
import makeAttachmentLoader from "./attachment";
import { makeAuditableLoader } from "./auditable";
import makeCustomerRequestedLanguageLoader from "./crl";
import { makeDescriptionLoader } from "./description";
import { makeDynamicStringLoader } from "./dynamicString";
import makeInvitationLoader from "./invitation";
import makeLanguageLoader from "./language";
import makeLocationLoader from "./location";
import {
  makeDisplayNameLoader,
  makeNameLoader,
  makeNameMetadataLoader,
} from "./name";
import makeOrganizationLoader from "./organization";
import { makeRequirementLoader } from "./requirement";
import { makeSopLoader } from "./sop";
import { makeStatusLoader } from "./status";
import makeTagLoader from "./tag";
import makeUserLoader from "./user";
import makeWorkerLoader from "./worker";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT: z.number({ coerce: true }).default(10),
  })
  .parse(process.env, { allowUnknown: true });

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  // TODO: this is probably a good idea...
  // connection: {
  //   statement_timeout: DB_STATEMENT_TIMEOUT * 1000, // milliseconds
  // },
});

export type SQL = typeof sql;
export type TxSQL = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}

export function orm(req: Request) {
  return {
    active: makeActiveLoader(req),
    attachment: makeAttachmentLoader(req),
    auditable: makeAuditableLoader(req),
    crl: makeCustomerRequestedLanguageLoader(req),
    description: makeDescriptionLoader(req),
    displayName: makeDisplayNameLoader(req),
    dynamicString: makeDynamicStringLoader(req),
    invitation: makeInvitationLoader(req),
    language: makeLanguageLoader(req),
    location: makeLocationLoader(req),
    name: makeNameLoader(req),
    nameMetadata: makeNameMetadataLoader(req),
    organization: makeOrganizationLoader(req),
    requirement: makeRequirementLoader(req),
    sop: makeSopLoader(req),
    status: makeStatusLoader(req),
    tag: makeTagLoader(req),
    user: makeUserLoader(req),
    worker: makeWorkerLoader(req),
  };
}

export type ORM = ReturnType<typeof orm>;
