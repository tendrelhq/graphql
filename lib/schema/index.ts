export * from "./__generated__/resolvers.generated";
export * from "./__generated__/typeDefs.generated";

export type Context = {
  authScope?: string;
  languageTypeId: number;
};
