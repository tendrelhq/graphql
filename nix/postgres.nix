{inputs, ...}: {
  imports = [
    inputs.process-compose.flakeModule
  ];
  perSystem = {
    config,
    self',
    pkgs,
    lib,
    ...
  }: {
    devShells.postgres = pkgs.mkShellNoCC {
      name = "tendrelhq/graphql+postgres";
      buildInputs = [
        config.packages.postgresql
        pkgs.perlPackages.TAPParserSourceHandlerpgTAP # pg_prove
        pkgs.python3.pkgs.sqlfmt
        pkgs.sqitchPg
        pkgs.squawk
      ];
    };

    packages = {
      inherit (pkgs) postgresql_14; # beta is on 14.12
      postgresql = pkgs.postgresql_16;
      pgddl = pkgs.stdenv.mkDerivation rec {
        name = "pgddl";
        version = "0.29";
        src = pkgs.fetchFromGitHub {
          owner = "lacanoid";
          repo = name;
          rev = version;
          hash = "sha256-W3G6TGtkj+zXXdGZZR0bmZhsLuFJvuGTlDoo8kL8sf0=";
        };
        buildInputs = with pkgs; [perl config.packages.postgresql];
        postPatch = ''
          patchShebangs .
        '';
        installPhase = ''
          install -D -t $out/share/postgresql/extension *.sql
          install -D -t $out/share/postgresql/extension *.control
        '';
      };
    };

    process-compose.devenv = {
      imports = [
        inputs.services.processComposeModules.default
      ];

      cli.options.no-server = false;

      services.postgres.pg1 = {
        enable = true;
        package = config.packages.postgresql;
        #
        extensions = exts:
          with exts;
          with config.packages; [
            pg_cron
            pgddl
            pgtap
            plpgsql_check
          ];
        hbaConf = lib.mkForce [
          # This one is the default docker bridge network.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.17.0.0/16";
            method = "trust";
          }
          # On my machine this is the tendrel_default network. YMMV.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.18.0.0/16";
            method = "trust";
          }
        ];
        initialDatabases = [
          {
            name = "dev";
          }
        ];
        settings = {
          listen_addresses = "0.0.0.0"; # so docker0 can connect
          log_statement = "all";
          logging_collector = false;
          shared_preload_libraries = "pg_cron,plpgsql_check";
        };
      };

      services.pgadmin.pga1 = {
        enable = true;
        extraConfig = {
          # Run in desktop mode with the master password disabled.
          MASTER_PASSWORD_REQUIRED = false;
          SERVER_MODE = false;
        };
        initialEmail = "";
        initialPassword = "";
      };

      settings.processes = {
        pga1 = {
          depends_on.pg1.condition = "process_healthy";
        };
      };
    };
  };
}
