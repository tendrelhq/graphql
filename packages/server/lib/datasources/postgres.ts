import postgres, { type Fragment } from "postgres";
import z from "zod";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT_MS } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT_MS: z.number({ coerce: true }).default(10_000),
  })
  .parse(process.env);

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  connection: {
    // We do some recursive queries and are, in general, not particularly
    // optimized so this is just a last resort to avoid bricking the backend.
    statement_timeout: DB_STATEMENT_TIMEOUT_MS,
  },
  types: {
    bigint: postgres.BigInt,
  },
});

export type Sql = typeof sql;
export type TxSql = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}
