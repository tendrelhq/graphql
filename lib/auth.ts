import { decodeGlobalId } from "@/schema/system";
import type { Context } from "@/schema/types";
import {
  ClerkExpressRequireAuth,
  type StrictAuthProp,
} from "@clerk/clerk-sdk-node";
import type e from "express";
import type { RequestHandler } from "express";
import { GraphQLError } from "graphql";
import * as jose from "jose";
import { type TxSql, sql } from "./datasources/postgres";
import { assert } from "./util";

declare global {
  namespace Express {
    interface Request extends StrictAuthProp {}
  }
}

export type Auth = StrictAuthProp["auth"];

export function clerk() {
  // In development it is nice to be able to use graphiql to help write your
  // queries. The default `devenv` will spin up a ruru[^1] graphiql endpoint
  // that you can use from your browser: http://localhost:1337.
  //
  // [^1]: https://github.com/graphile/crystal/tree/main/grafast/ruru
  //
  // tl;dr if you set the 'x-tendrel-user' header to a Clerk user id, we will
  // use that as the authenticated identity
  return (req: e.Request, res: e.Response, next: e.NextFunction) => {
    if (process.env.NODE_ENV === "development") {
      const userId = req.headers["x-tendrel-user"];
      console.log(`Backdoor hack engaged? ${userId}`);
      if (userId) {
        req.auth = {
          userId: userId as string,
        } as Auth;

        return next();
      }
    }

    ClerkExpressRequireAuth()(req, res, err => {
      if (err instanceof Error) {
        // @ts-ignore
        err.statusCode = 401;
      }

      next(err);
    });
  };
}

/**
 * Verify the provided *external* security token, generate a short-lived,
 * finely-grained subject token, and then call our /token api to perform the
 * actual token exchange. The client should not be aware that we are performing
 * this multi-step process.
 */
export const login: RequestHandler = async (req, res) => {
  // TODO: Move this elsewhere.
  const ALG = process.env.JWT_ALG ?? "HS256";
  const SECRET = new TextEncoder().encode(process.env.JWT_SECRET);
  // N.B. when JWT_ALG = HS256, JWT_SECRET must be at least 32 bytes!
  if (ALG === "HS256") assert(SECRET.length >= 32, "invalid jwt secret");

  const { userId } = req.auth;

  // TODO: ensure the user exists in the system, effectively making this API a
  // dual login/signup API. For now we will assume that the user has already
  // "signed up" for a Tendrel account, i.e. via the canonical Clerk/console flow.

  const temp = await new jose.SignJWT({ role: "anonymous" })
    .setProtectedHeader({ alg: ALG, typ: "JWT" })
    .setIssuer(`urn:tendrel:${process.env.STAGE}`)
    .setIssuedAt() // now
    .setNotBefore("-5 minutes")
    .setExpirationTime("5 minutes")
    .setSubject(userId)
    .sign(SECRET);

  try {
    // Same URL in both development and production. In development this will hit
    // the nginx proxy running on localhost. In production it will hit the same
    // proxy but running in a separate (ECS) container.
    const token = await fetch("http://localhost/api/v1/rpc/token", {
      method: "POST",
      body: JSON.stringify({
        grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
        subject_token: temp,
        subject_token_type: "urn:ietf:params:oauth:token-type:jwt",
      }),
      headers: {
        Authorization: `Bearer ${temp}`,
      },
    });

    const json = await token.json();

    res
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.1
      .setHeader("Cache-Control", "no-store")
      .setHeader("Pragma", "no-cache")
      .status(token.status);

    if (!token.ok) {
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.2
      res.json({
        error: json.code,
        error_description: json.message,
      });
    } else {
      res.json(json);
    }
  } catch (e) {
    console.error(e);
    res.status(500).send("Internal Server Error");
  }
};

export async function protect(
  { orgId, userId }: { orgId: string; userId: string },
  allowed: ("Worker" | "Supervisor" | "Admin")[],
) {
  const rows = await sql`
    select 1
    from public.workerinstance as w
    inner join public.systag as s on w.workerinstanceuserroleid = s.systagid
    where
        w.workerinstancecustomerid = (
            select customerid
            from public.customer
            where customeruuid = ${decodeGlobalId(orgId).id}
        )
        and w.workerinstanceworkerid = (
            select workerid
            from public.worker
            where workeridentityid = ${userId}
        )
        and s.systagtype in (${sql(allowed)});
  `;

  if (!rows.length) {
    throw new GraphQLError("Not authorized", {
      extensions: {
        code: "UNAUTHORIZED",
        hint: "You do not have the necessary permissions to perform this action",
      },
    });
  }
}

export function setCurrentIdentity(sql: TxSql, ctx: Context) {
  return sql`select * from auth.set_actor(${ctx.auth.userId}, ${ctx.req.i18n.language})`;
}
