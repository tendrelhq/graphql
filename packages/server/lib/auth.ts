import { decodeGlobalId } from "@/schema/system";
import type { Context } from "@/schema/types";
import {
  ClerkExpressRequireAuth,
  type StrictAuthProp,
} from "@clerk/clerk-sdk-node";
import type e from "express";
import type { RequestHandler } from "express";
import { GraphQLError } from "graphql";
import config from "./config";
import { type TxSql, sql } from "./datasources/postgres";

declare global {
  namespace Express {
    interface Request extends StrictAuthProp {}
  }
}

export type Auth = StrictAuthProp["auth"];

export function clerk() {
  // In development it is nice to be able to use graphiql to help write your
  // queries. The default `devenv` will spin up a ruru[^1] graphiql endpoint
  // that you can use from your browser: http://localhost:1337.
  //
  // [^1]: https://github.com/graphile/crystal/tree/main/grafast/ruru
  //
  // tl;dr if you set the 'x-tendrel-user' header to a Clerk user id, we will
  // use that as the authenticated identity
  return (req: e.Request, res: e.Response, next: e.NextFunction) => {
    if (process.env.NODE_ENV === "development") {
      const userId = req.headers["x-tendrel-user"];
      console.log(`Backdoor hack engaged? ${userId}`);
      if (userId) {
        req.auth = {
          userId: userId as string,
        } as Auth;

        return next();
      }
    }

    ClerkExpressRequireAuth()(req, res, err => {
      if (err instanceof Error) {
        // @ts-ignore
        err.statusCode = 401;
      }

      next(err);
    });
  };
}

/**
 * Verify the provided *external* security token, generate a short-lived,
 * finely-grained subject token, and then call our /token api to perform the
 * actual token exchange. The client should not be aware that we are performing
 * this multi-step process.
 */
export const login: RequestHandler = async (req, res) => {
  try {
    const r = await getAccessToken(req.auth.userId);
    const token = await r.json();

    res
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.1
      .setHeader("Cache-Control", "no-store")
      .setHeader("Pragma", "no-cache")
      .status(r.status);

    if (!r.ok) {
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.2
      res.json({
        error: token.code,
        error_description: token.message,
      });
    } else {
      res.json(token);
    }
  } catch (e) {
    console.error(e);
    res.status(500).send("Internal Server Error");
  }
};

const ACCESS_TOKEN_ENDPOINT = new URL("/api/v1/rpc/token", config.base_url);

/**
 * Get an access token for the given `sub`ject. This is always a Clerk user ID
 * at the moment.
 */
export async function getAccessToken(sub: string) {
  const [{ jwt }] = await sql`
    select auth.jwt_sign(
      json_build_object(
        'role', 'anonymous',
        'iss', ${config.jwt_iss}::text,
        'iat', extract(epoch from now()),
        'nbf', extract(epoch from now()) - 30 /* seconds */,
        'exp', extract(epoch from now()) + 30 /* seconds */,
        'sub', ${sub}::text
      )
    ) as jwt;
  `;

  // Same URL in both development and production. In development this assumes
  // that Postgrest is running locally on port 4001, which it will be if you
  // use the docker-compose file.
  return await fetch(ACCESS_TOKEN_ENDPOINT, {
    method: "POST",
    body: JSON.stringify({
      grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
      subject_token: jwt,
      subject_token_type: "urn:ietf:params:oauth:token-type:jwt",
    }),
    headers: {
      Authorization: `Bearer ${jwt}`,
    },
  });
}

export async function protect(
  { orgId, userId }: { orgId: string; userId: string },
  allowed: ("Worker" | "Supervisor" | "Admin")[],
) {
  const rows = await sql`
    select 1
    from public.workerinstance as w
    inner join public.systag as s on w.workerinstanceuserroleid = s.systagid
    where
        w.workerinstancecustomerid = (
            select customerid
            from public.customer
            where customeruuid = ${decodeGlobalId(orgId).id}
        )
        and w.workerinstanceworkerid = (
            select workerid
            from public.worker
            where workeridentityid = ${userId}
        )
        and s.systagtype in (${sql(allowed)});
  `;

  if (!rows.length) {
    throw new GraphQLError("Not authorized", {
      extensions: {
        code: "UNAUTHORIZED",
        hint: "You do not have the necessary permissions to perform this action",
      },
    });
  }
}

export function setCurrentIdentity(sql: TxSql, ctx: Context) {
  return sql`select * from auth.set_actor(${ctx.auth.userId}, ${ctx.req.i18n.language})`;
}
