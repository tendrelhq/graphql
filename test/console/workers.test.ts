import { schema } from "@/schema/final";
import {
  cleanup,
  createTestContext,
  execute,
  findAndEncode,
  paginateQuery,
  setup,
} from "@/test/prelude";
import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import {
  ListWorkersTestDocument,
  PaginateWorkersTestDocument,
} from "./workers.test.generated";
import { setCurrentIdentity } from "@/auth";
import { sql } from "@/datasources/postgres";

const ctx = await createTestContext();

describe("[console] workers", () => {
  let CUSTOMER: string;

  test("list", async () => {
    const result = await execute(schema, ListWorkersTestDocument, {
      account: CUSTOMER,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("paginate", async () => {
    let i = 0;
    for await (const page of paginateQuery({
      async execute(cursor) {
        return await execute(schema, PaginateWorkersTestDocument, {
          account: CUSTOMER,
          first: 5,
          after: cursor,
        });
      },
      next(result) {
        if (result.data?.node.__typename !== "Organization") {
          throw "invariant violated";
        }
        return result.data.node?.workers.pageInfo;
      },
    })) {
      expect(page.errors).toBeFalsy();
      i++;
    }
    expect(i).toBe(3); // 12 total, 5 per page
  });

  test("search", async () => {
    const result = await execute(schema, ListWorkersTestDocument, {
      account: CUSTOMER,
      search: {
        active: true,
        user: {
          displayName: "will",
        },
      },
    });
    expect(result.errors).toBeFalsy();
    expect(result.data?.node.__typename).toBe("Organization");
    if (result.data?.node.__typename === "Organization") {
      expect(result.data.node.workers.totalCount).toBe(8);
    }
  });

  beforeAll(async () => {
    const logs = await sql.begin(async sql => {
      await setCurrentIdentity(sql, ctx);
      return await sql<{ op: string; id: string }[]>`
        select *
        from
            runtime.create_demo(
                customer_name := 'Frozen Tendy Factory',
                admins := (
                    select array_agg(workeruuid)
                    from public.worker
                    where workerfullname in (
                        'Jerry Garcia',
                        'Mike Heavner',
                        'Will Ruggiano',
                        'Will Twait'
                    )
                ),
                modified_by := 895
            )
        ;
      `;
    });
    CUSTOMER = findAndEncode("customer", "organization", logs);
  });

  afterAll(async () => {
    await cleanup(CUSTOMER);
  });
});
