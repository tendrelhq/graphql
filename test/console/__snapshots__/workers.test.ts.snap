// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`[console] workers list 1`] = `
{
  "node": {
    "__typename": "Organization",
    "workers": {
      "__typename": "WorkerConnection",
      "edges": [
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Jerry Garcia",
            "firstName": "Jerry",
            "language": {
              "code": "en",
            },
            "lastName": "Garcia",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Jerry Garcia",
            "firstName": "Jerry",
            "language": {
              "code": "en",
            },
            "lastName": "Garcia",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Jerry Garcia",
            "firstName": "Jerry",
            "language": {
              "code": "en",
            },
            "lastName": "Garcia",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Mike Heavner",
            "firstName": "Mike",
            "language": {
              "code": "en",
            },
            "lastName": "Heavner",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Mike Heavner",
            "firstName": "Mike",
            "language": {
              "code": "en",
            },
            "lastName": "Heavner",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Mike Heavner",
            "firstName": "Mike",
            "language": {
              "code": "en",
            },
            "lastName": "Heavner",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Will Ruggiano",
            "firstName": "Will",
            "language": {
              "code": "en",
            },
            "lastName": "Ruggiano",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
        {
          "__typename": "WorkerEdge",
          "node": {
            "__typename": "Worker",
            "active": {
              "active": true,
            },
            "displayName": "Will Ruggiano",
            "firstName": "Will",
            "language": {
              "code": "en",
            },
            "lastName": "Ruggiano",
            "role": {
              "type": "Admin",
            },
            "scanCode": null,
          },
        },
      ],
      "pageInfo": {
        "__typename": "PageInfo",
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 8,
    },
  },
}
`;
